llvm/lib/CodeGen/SelectOptimize.cpp

```diff
@@ -749,93 +717,160 @@ void SelectOptimizeImpl::convertProfitableSIGroups(SelectGroups &ProfSIGroups) {
-    for (auto SI : ASI)
+    for (SelectLike &SI : ASI.Selects)
       SI.getI()->eraseFromParent();
   }
 }
 
 void SelectOptimizeImpl::collectSelectGroups(BasicBlock &BB,
                                              SelectGroups &SIGroups) {
+  // Represents something that can be considered as select instruction.
+  // Auxiliary instruction are instructions that depends on a condition and have
+  // zero or some constant value on True/False branch, such as:
+  // * ZExt(1bit), SExt(1bit)
+  // * Not(1bit)
+  // * AShr(Xbit), X-1, LShr(XBit), X-1, where there is a condition like Xbit <=
+  // 0 somewhere above in BB
+  struct SelectLikeInfo {
+    Value *Cond;
+    bool IsAuxiliary;
+    bool IsInverted;
+    unsigned ConditionIdx;
+  };
+
+  SmallPtrSet<Instruction *, 2> SeenCmp;
+  std::map<Value *, SelectLikeInfo> SelectInfo;
+
   BasicBlock::iterator BBIt = BB.begin();
   while (BBIt != BB.end()) {
     Instruction *I = &*BBIt++;
-    if (SelectLike SI = SelectLike::match(I)) {
-      if (!TTI->shouldTreatInstructionLikeSelect(I))
+    if (auto *Cmp = dyn_cast<CmpInst>(I)) {
+      SeenCmp.insert(Cmp);
+      continue;
+    }
+
+    Value *Cond;
+    if (match(I, m_OneUse(m_ZExt(m_Value(Cond)))) &&
+        Cond->getType()->isIntegerTy(1)) {
+      bool Inverted = match(Cond, m_Not(m_Value(Cond)));
+      SelectInfo[I] = {Cond, true, Inverted, 0};
+      continue;
+    }
+
+    if (match(I, m_Not(m_Value(Cond)))) {
+      SelectInfo[I] = {Cond, true, true, 0};
+      continue;
+    }
+
+    // Select instruction are what we are usually looking for.
+    if (match(I, m_Select(m_Value(Cond), m_Value(), m_Value()))) {
+      bool Inverted = match(Cond, m_Not(m_Value(Cond)));
+      SelectInfo[I] = {Cond, false, Inverted, 0};
+      continue;
+    }
+
+    // An Or(zext(i1 X), Y) can also be treated like a select, with condition
+    if (auto *BO = dyn_cast<BinaryOperator>(I)) {
+      if (BO->getType()->isIntegerTy(1) || BO->getOpcode() != Instruction::Or)
         continue;
 
-      SelectGroup SIGroup;
-      SIGroup.push_back(SI);
-      while (BBIt != BB.end()) {
-        Instruction *NI = &*BBIt;
-        // Debug/pseudo instructions should be skipped and not prevent the
-        // formation of a select group.
-        if (NI->isDebugOrPseudoInst()) {
-          ++BBIt;
-          continue;
+      for (unsigned Idx = 0; Idx < 2; Idx++) {
+        auto *Op = BO->getOperand(Idx);
+        if (SelectInfo.count(Op) && SelectInfo[Op].IsAuxiliary) {
+          Cond = SelectInfo[Op].Cond;
+          bool Inverted = SelectInfo[Op].IsInverted;
+          SelectInfo[I] = {Cond, false, Inverted, Idx};
+          break;
         }
+      }
+      continue;
+    }
+  }
 
-        // Skip not(select(..)), if the not is part of the same select group
-        if (match(NI, m_Not(m_Specific(SI.getCondition())))) {
-          ++BBIt;
-          continue;
-        }
+  BBIt = BB.begin();
+  while (BBIt != BB.end()) {
```

Comment:
Is it possible to avoid the second pass of instructions, maybe by moving the recognition of select info in the loop above into a lambda that could be used in this loop?

Reply:
Something like this?

Reply:
Yeah that sounds good. I was wondering if there was a way to remove the SelectInfo map entirely and return an optional SelectLikeInfo. Just incase there are very large functions with many different select instructions in them. It might involve re-checking some instructions, but that might not be worse than putting them into a map.
I'm not sure if that would make what you are adding in the future more difficult, Up to you whether that sounds like an improvement or not.

Reply:
What do you think about this way? Now, we'll have to access the map only when processing the Or/Later Add/Sub operation.